{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Untitled9.ipynb",
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "**Подготавливаем файлы для аннотации**"
      ],
      "metadata": {
        "id": "2LGC9N4xvNwl"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Скачиваем собранный геном бактерии (все скаффолды)"
      ],
      "metadata": {
        "id": "-fH9a2DMvTy5"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://raw.githubusercontent.com/alexburde/hse21_hw1/main/data/Poil_scaffold.fa"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "9C_M8MONvPp4",
        "outputId": "7effd798-7275-4070-eb25-035bdcbe5e10"
      },
      "execution_count": 31,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-12-17 16:00:44--  https://raw.githubusercontent.com/alexburde/hse21_hw1/main/data/Poil_scaffold.fa\n",
            "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 3926990 (3.7M) [text/plain]\n",
            "Saving to: ‘Poil_scaffold.fa’\n",
            "\n",
            "Poil_scaffold.fa    100%[===================>]   3.74M  --.-KB/s    in 0.07s   \n",
            "\n",
            "2021-12-17 16:00:44 (53.6 MB/s) - ‘Poil_scaffold.fa’ saved [3926990/3926990]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Установка GeneMarkS-2**"
      ],
      "metadata": {
        "id": "zae2FcLu6YUE"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget http://topaz.gatech.edu/GeneMark/tmp/GMtool_lZ6Ul/gms2_linux_64.tar.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "fKwXhI-Q6F-k",
        "outputId": "6aa36856-d82a-4c03-e2b7-dc60a68e8301"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-12-17 15:07:39--  http://topaz.gatech.edu/GeneMark/tmp/GMtool_lZ6Ul/gms2_linux_64.tar.gz\n",
            "Resolving topaz.gatech.edu (topaz.gatech.edu)... 143.215.161.100\n",
            "Connecting to topaz.gatech.edu (topaz.gatech.edu)|143.215.161.100|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 11261783 (11M) [application/x-gzip]\n",
            "Saving to: ‘gms2_linux_64.tar.gz’\n",
            "\n",
            "gms2_linux_64.tar.g 100%[===================>]  10.74M  5.29MB/s    in 2.0s    \n",
            "\n",
            "2021-12-17 15:07:42 (5.29 MB/s) - ‘gms2_linux_64.tar.gz’ saved [11261783/11261783]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!wget http://topaz.gatech.edu/GeneMark/tmp/GMtool_lZ6Ul/gm_key_64.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "U7v5fQSH71HV",
        "outputId": "0a11d304-d7b0-457d-a4f3-bbc8bb9e950b"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-12-17 15:08:16--  http://topaz.gatech.edu/GeneMark/tmp/GMtool_lZ6Ul/gm_key_64.gz\n",
            "Resolving topaz.gatech.edu (topaz.gatech.edu)... 143.215.161.100\n",
            "Connecting to topaz.gatech.edu (topaz.gatech.edu)|143.215.161.100|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 86 [application/x-gzip]\n",
            "Saving to: ‘gm_key_64.gz’\n",
            "\n",
            "\rgm_key_64.gz          0%[                    ]       0  --.-KB/s               \rgm_key_64.gz        100%[===================>]      86  --.-KB/s    in 0s      \n",
            "\n",
            "2021-12-17 15:08:16 (10.9 MB/s) - ‘gm_key_64.gz’ saved [86/86]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!gzip -d gm_key_64.gz"
      ],
      "metadata": {
        "id": "UqEhKMAN8ATF"
      },
      "execution_count": 8,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!tar -xzvf gms2_linux_64.tar.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "koMU8OWS8H-l",
        "outputId": "cec9c057-b781-42a5-b860-cd6da3e626ee"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "gms2_linux_64/\n",
            "gms2_linux_64/gms2.pl\n",
            "gms2_linux_64/README_GMS2.txt\n",
            "gms2_linux_64/gmhmmp2_key\n",
            "gms2_linux_64/mgm_4.mod\n",
            "gms2_linux_64/gmhmmp2\n",
            "gms2_linux_64/INSTALL\n",
            "gms2_linux_64/mgm_11.mod\n",
            "gms2_linux_64/biogem\n",
            "gms2_linux_64/LICENSE\n",
            "gms2_linux_64/mgm_25.mod\n",
            "gms2_linux_64/compp\n",
            "gms2_linux_64/mgm_15.mod\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!cp  -v  gm_key_64   ~/.gmhmmp2_key"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0m6sOAOE8IFY",
        "outputId": "a99992e8-ba78-4eea-9a49-9adf106345aa"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "'gm_key_64' -> '/root/.gmhmmp2_key'\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Установка E-utilities (для скачивания последовательностей из NCBI)**"
      ],
      "metadata": {
        "id": "ZMGxjOrr8VO7"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!sh -c \"$(curl -fsSL ftp://ftp.ncbi.nlm.nih.gov/entrez/entrezdirect/install-edirect.sh)\""
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "d-rp9_hk8IHu",
        "outputId": "38a7c168-600d-4207-b92d-b6d287e44575"
      },
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Entrez Direct has been successfully downloaded and installed.\n",
            "\n",
            "In order to complete the configuration process, please execute the following:\n",
            "\n",
            "  echo \"export PATH=\\${PATH}:/root/edirect\" >> ${HOME}/.bashrc\n",
            "\n",
            "or manually edit the PATH variable assignment in your .bashrc file.\n",
            "\n",
            "Would you like to do that automatically now? [y/N]\n",
            "y\n",
            "OK, done.\n",
            "\n",
            "To activate EDirect for this terminal session, please execute the following:\n",
            "\n",
            "export PATH=${PATH}:${HOME}/edirect\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!$HOME/edirect/efetch -db nuccore -id HF680312 -format gb  >  T_oleivorans_MIL_1.gbk"
      ],
      "metadata": {
        "id": "HjWxn2fq8IJr"
      },
      "execution_count": 12,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!$HOME/edirect/efetch -db nuccore -id HF680312 -format gene_fasta  >  T_oleivorans_MIL_1.genes.fasta"
      ],
      "metadata": {
        "id": "nvYlSb1u87l_"
      },
      "execution_count": 13,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!$HOME/edirect/efetch -db nuccore -id HF680312 -format fasta_cds_aa  >  T_oleivorans_MIL_1.proteins.fasta"
      ],
      "metadata": {
        "id": "5pjTvKUz87qV"
      },
      "execution_count": 14,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Установка BLAST**"
      ],
      "metadata": {
        "id": "LgVS00Xm9bBP"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!apt-get update"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "MT9fH3hq9drR",
        "outputId": "28bda9b9-364e-4b10-f6dc-d6495f1f5298"
      },
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\r0% [Working]\r            \rGet:1 http://security.ubuntu.com/ubuntu bionic-security InRelease [88.7 kB]\n",
            "\r0% [Connecting to archive.ubuntu.com (91.189.88.152)] [1 InRelease 14.2 kB/88.7\r                                                                               \rGet:2 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ InRelease [3,626 B]\n",
            "Ign:3 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease\n",
            "Get:4 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic InRelease [15.9 kB]\n",
            "Ign:5 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  InRelease\n",
            "Get:6 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Release [696 B]\n",
            "Hit:7 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  Release\n",
            "Get:8 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Release.gpg [836 B]\n",
            "Hit:9 http://archive.ubuntu.com/ubuntu bionic InRelease\n",
            "Get:10 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ Packages [73.9 kB]\n",
            "Get:11 http://archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB]\n",
            "Hit:12 http://ppa.launchpad.net/cran/libgit2/ubuntu bionic InRelease\n",
            "Get:13 http://security.ubuntu.com/ubuntu bionic-security/main amd64 Packages [2,461 kB]\n",
            "Get:14 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu bionic InRelease [15.9 kB]\n",
            "Get:15 http://archive.ubuntu.com/ubuntu bionic-backports InRelease [74.6 kB]\n",
            "Get:17 http://security.ubuntu.com/ubuntu bionic-security/restricted amd64 Packages [691 kB]\n",
            "Get:18 http://security.ubuntu.com/ubuntu bionic-security/universe amd64 Packages [1,452 kB]\n",
            "Hit:19 http://ppa.launchpad.net/graphics-drivers/ppa/ubuntu bionic InRelease\n",
            "Get:20 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Packages [833 kB]\n",
            "Get:21 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic/main Sources [1,820 kB]\n",
            "Get:22 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 Packages [2,898 kB]\n",
            "Get:23 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic/main amd64 Packages [933 kB]\n",
            "Get:24 http://archive.ubuntu.com/ubuntu bionic-updates/universe amd64 Packages [2,230 kB]\n",
            "Get:25 http://archive.ubuntu.com/ubuntu bionic-backports/universe amd64 Packages [12.6 kB]\n",
            "Get:26 http://archive.ubuntu.com/ubuntu bionic-backports/main amd64 Packages [11.6 kB]\n",
            "Get:27 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu bionic/main amd64 Packages [45.9 kB]\n",
            "Fetched 13.8 MB in 4s (3,609 kB/s)\n",
            "Reading package lists... Done\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!apt-get install ncbi-blast+"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ST_U8Jub9hW6",
        "outputId": "213c5541-8559-4a4e-817d-90d01c340b1a"
      },
      "execution_count": 16,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  ncbi-data\n",
            "The following NEW packages will be installed:\n",
            "  ncbi-blast+ ncbi-data\n",
            "0 upgraded, 2 newly installed, 0 to remove and 79 not upgraded.\n",
            "Need to get 13.1 MB of archives.\n",
            "After this operation, 66.7 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 ncbi-data all 6.1.20170106-2 [3,645 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu bionic/universe amd64 ncbi-blast+ amd64 2.6.0-1 [9,446 kB]\n",
            "Fetched 13.1 MB in 2s (8,220 kB/s)\n",
            "Selecting previously unselected package ncbi-data.\n",
            "(Reading database ... 155222 files and directories currently installed.)\n",
            "Preparing to unpack .../ncbi-data_6.1.20170106-2_all.deb ...\n",
            "Unpacking ncbi-data (6.1.20170106-2) ...\n",
            "Selecting previously unselected package ncbi-blast+.\n",
            "Preparing to unpack .../ncbi-blast+_2.6.0-1_amd64.deb ...\n",
            "Unpacking ncbi-blast+ (2.6.0-1) ...\n",
            "Setting up ncbi-data (6.1.20170106-2) ...\n",
            "Setting up ncbi-blast+ (2.6.0-1) ...\n",
            "Processing triggers for hicolor-icon-theme (0.17-2) ...\n",
            "Processing triggers for man-db (2.8.3-2ubuntu0.1) ...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Скачиваем БД белков SwissProt**"
      ],
      "metadata": {
        "id": "aoG9IeNk9uNy"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "NCnZFTbf9v1y",
        "outputId": "211ca244-ce59-41d5-846b-0cb38e4a7ecf"
      },
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-12-17 15:17:14--  https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz\n",
            "Resolving ftp.uniprot.org (ftp.uniprot.org)... 128.175.240.195\n",
            "Connecting to ftp.uniprot.org (ftp.uniprot.org)|128.175.240.195|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 90527596 (86M) [application/x-gzip]\n",
            "Saving to: ‘uniprot_sprot.fasta.gz’\n",
            "\n",
            "uniprot_sprot.fasta 100%[===================>]  86.33M  70.9MB/s    in 1.2s    \n",
            "\n",
            "2021-12-17 15:17:15 (70.9 MB/s) - ‘uniprot_sprot.fasta.gz’ saved [90527596/90527596]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!gzip -d uniprot_sprot.fasta.gz"
      ],
      "metadata": {
        "id": "_X6PXoI6-AMU"
      },
      "execution_count": 18,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Устанавливаем пакет seqkit**"
      ],
      "metadata": {
        "id": "EAIW1u7m-EyG"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://github.com/shenwei356/seqkit/releases/download/v2.1.0/seqkit_linux_amd64.tar.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2KKp1Lnc-F7v",
        "outputId": "4471f5f1-3b31-409d-b5c0-3401fe8cd3f6"
      },
      "execution_count": 19,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2021-12-17 15:18:25--  https://github.com/shenwei356/seqkit/releases/download/v2.1.0/seqkit_linux_amd64.tar.gz\n",
            "Resolving github.com (github.com)... 140.82.112.3\n",
            "Connecting to github.com (github.com)|140.82.112.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/52715040/ea4018ec-70e8-40ac-a4aa-3326b04e07c4?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20211217%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20211217T151825Z&X-Amz-Expires=300&X-Amz-Signature=6e1f57d58fadb99d945b71b22544d8de3518db24e2fb528fced265ed1fc2ed6e&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=52715040&response-content-disposition=attachment%3B%20filename%3Dseqkit_linux_amd64.tar.gz&response-content-type=application%2Foctet-stream [following]\n",
            "--2021-12-17 15:18:25--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/52715040/ea4018ec-70e8-40ac-a4aa-3326b04e07c4?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20211217%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20211217T151825Z&X-Amz-Expires=300&X-Amz-Signature=6e1f57d58fadb99d945b71b22544d8de3518db24e2fb528fced265ed1fc2ed6e&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=52715040&response-content-disposition=attachment%3B%20filename%3Dseqkit_linux_amd64.tar.gz&response-content-type=application%2Foctet-stream\n",
            "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n",
            "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.108.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 6650850 (6.3M) [application/octet-stream]\n",
            "Saving to: ‘seqkit_linux_amd64.tar.gz’\n",
            "\n",
            "seqkit_linux_amd64. 100%[===================>]   6.34M  --.-KB/s    in 0.08s   \n",
            "\n",
            "2021-12-17 15:18:26 (75.3 MB/s) - ‘seqkit_linux_amd64.tar.gz’ saved [6650850/6650850]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!tar -xzvf seqkit_linux_amd64.tar.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "T5PFlo7X-SLz",
        "outputId": "ff1667b1-cd63-47fa-ae76-0812a6a4e894"
      },
      "execution_count": 20,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "seqkit\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!chmod a+x seqkit"
      ],
      "metadata": {
        "id": "Eb1MlGSZ-WUc"
      },
      "execution_count": 21,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Выполнение задания**"
      ],
      "metadata": {
        "id": "XXtXm6pV-fl3"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Предсказать расположение белок-кодирующих генов в геноме нашей бактерии**"
      ],
      "metadata": {
        "id": "4J7XXS-N-lr-"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Предсказать белок-кодирующие гены в собранном геноме нашей бактерии с помощью ab initio программы GeneMarkS – комманда gms2.pl\n",
        "По умолчанию координаты предсказанных генов сохраняются в файле gms2.lst\n",
        "Обратите внимание, что GeneMarkS может выдавать последовательности генов (опция --fnn), а также автоматически их транслировать в белки (--faa)"
      ],
      "metadata": {
        "id": "mWwiwjAr-qFV"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!./gms2_linux_64/gms2.pl --seq Poil_scaffold.fa --genome-type bacteria --fnn genes.fasta --faa proteins.fasta"
      ],
      "metadata": {
        "id": "5n47k2G5AoMp"
      },
      "execution_count": 32,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Определяем функции белков нашей бактерии через сравнение с белками из T.oleivorans MIL-1**"
      ],
      "metadata": {
        "id": "zclMnAM0BXlQ"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Сравнить все белки (протеом) нашей бактерии с аннотированными белками T. oleivorans:\n",
        "\n",
        "Протеом T. oleivorans MIL-1 находится в файле T_oleivorans_MIL_1.proteins.fasta\n",
        "\n",
        "Индексируем белки T.oleivorans MIL-1:"
      ],
      "metadata": {
        "id": "IzgVnNi3BZVz"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!makeblastdb  -dbtype prot  -in T_oleivorans_MIL_1.proteins.fasta  -out T_oleivorans_MIL_1.proteins"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ys6E4dA1Bb6B",
        "outputId": "154d858e-e5b7-40ee-ad72-3eee6c6fb68b"
      },
      "execution_count": 33,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "\n",
            "Building a new DB, current time: 12/17/2021 16:05:59\n",
            "New DB name:   /content/T_oleivorans_MIL_1.proteins\n",
            "New DB title:  T_oleivorans_MIL_1.proteins.fasta\n",
            "Sequence type: Protein\n",
            "Keep MBits: T\n",
            "Maximum file size: 1000000000B\n",
            "Adding sequences from FASTA; added 3666 sequences in 0.113555 seconds.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Определеям какие белки из нашей бактерии похожи на белки из T_oleivorans:"
      ],
      "metadata": {
        "id": "TTTwNxjNBr0f"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!blastp  -query proteins.fasta  -db T_oleivorans_MIL_1.proteins  -evalue 1e-10  -outfmt 6  >  scaffolds.hits_from_MIL_1.txt"
      ],
      "metadata": {
        "id": "2HzwB8emB341"
      },
      "execution_count": 34,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Определяем функции белков нашей бактерии через поиск по БД SwissProt**"
      ],
      "metadata": {
        "id": "FQBJ_fBMB9Sa"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Смотрим, какие белки нашей бактерии (из Баренцева моря) не имели похожих белков из бактерии T.oleivorans MIL-1\n",
        "\n",
        "Для этого получаем список уникальных номеров генов бактерии из Барнцева моря, коротые имели хотя бы один похожий белок бактерии из Сицилли"
      ],
      "metadata": {
        "id": "dvSVBjDfCFk4"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!cut -f 1 scaffolds.hits_from_MIL_1.txt | sort -n | uniq > proteins.with_hits_from_MIL_1.txt"
      ],
      "metadata": {
        "id": "nh8zgUVzCA8B"
      },
      "execution_count": 35,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!./seqkit grep  --invert-match  -f proteins.with_hits_from_MIL_1.txt  proteins.fasta   -o proteins.without_MIL_1.fasta"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "tlzBbvb3CJjG",
        "outputId": "33a13de0-051e-4400-a923-0989902379d4"
      },
      "execution_count": 36,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[INFO]\u001b[0m 3278 patterns loaded from file\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Проверяем, что кол-ва белков сходятся\n",
        "\n",
        "Общее кол-во белков в нашей бактерии:"
      ],
      "metadata": {
        "id": "Gzp6ejJnCOfj"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!grep '>' proteins.fasta | wc -l"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "88a6CAOrCMC3",
        "outputId": "51fe0c1d-c2df-4157-a766-52a6c079407e"
      },
      "execution_count": 37,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "3567\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Кол-во белков, которые имели схожие белки из бактерии MIL-1"
      ],
      "metadata": {
        "id": "3hYI9Ps8CYE3"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wc -l proteins.with_hits_from_MIL_1.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2QRECG6xCZwK",
        "outputId": "a5ad5c7e-2a00-4374-eb7d-9914c7b1eb3c"
      },
      "execution_count": 38,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "3278 proteins.with_hits_from_MIL_1.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!grep '>' proteins.without_MIL_1.fasta | wc -l"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "HYabqheICdxu",
        "outputId": "5c5af9f4-6c3f-408d-da93-77bc4f18e423"
      },
      "execution_count": 39,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "289\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Сравнить все белки (протеом) нашей бактерии с аннотированными белками из БД SwissProt\n",
        "\n",
        "Индексируем белки SwissProt:"
      ],
      "metadata": {
        "id": "6iH5WC0ECYuE"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!makeblastdb  -dbtype prot  -in uniprot_sprot.fasta  -out uniprot_sprot"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "l-DDS-0SCT01",
        "outputId": "fb542c86-9955-4b84-bdf0-01a393ac82cd"
      },
      "execution_count": 40,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "\n",
            "Building a new DB, current time: 12/17/2021 16:18:11\n",
            "New DB name:   /content/uniprot_sprot\n",
            "New DB title:  uniprot_sprot.fasta\n",
            "Sequence type: Protein\n",
            "Keep MBits: T\n",
            "Maximum file size: 1000000000B\n",
            "Adding sequences from FASTA; added 565928 sequences in 14.514 seconds.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Определеям какие белки из нашей бактерии похожи на белки из БД SwissProt (программа работает около 10 мин):"
      ],
      "metadata": {
        "id": "eYO5OrlcCm90"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!time  blastp  -query proteins.without_MIL_1.fasta  -db uniprot_sprot  -evalue 1e-10  -outfmt 6  >  scaffolds.hits_from_SwissProt.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "hTy85TXhCknQ",
        "outputId": "231163a8-1a7b-4b15-86f7-46028102cf76"
      },
      "execution_count": 41,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "real\t9m5.111s\n",
            "user\t9m4.059s\n",
            "sys\t0m0.389s\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Посмотрим, сколько белков удалось аннотировать через поиск по SwissProt:"
      ],
      "metadata": {
        "id": "lrWyTL1uCsOx"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!cut -f 1 scaffolds.hits_from_SwissProt.txt | sort | uniq | wc -l"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "dltBj_s0CqGV",
        "outputId": "dbb788d7-86cc-40ba-84f5-41b0e51bc575"
      },
      "execution_count": 45,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "52\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Загружаем файлы, которые нам потребуются для аннотации генома, на github"
      ],
      "metadata": {
        "id": "8nmoYdb8Cu9O"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "gms2.lst -- координаты всех предсказанных генов\n",
        "\n",
        "proteins.fasta -- аминокислотные последовательности всех предсказанных генов\n",
        "\n",
        "scaffolds.hits_from_MIL_1.txt -- информация о схожести белков из нашей бактерии с белками из бактерии MIL-1\n",
        "\n",
        "scaffolds.hits_from_SwissProt.txt -- информация о схожести белков из нашей бактерии с белками из БД SwissProt"
      ],
      "metadata": {
        "id": "S4hstyK9Cy0A"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Дальше пишем код на Python с использованием библиотеки Biopython и создаем геном с его аннотацией в формате GenBank (можно сделать в отдельном ноутбуке)."
      ],
      "metadata": {
        "id": "8hIXFgCrC1C4"
      }
    }
  ]
}